be/src/connector/connector.h

```diff
@@ -8,56 +8,77 @@
 
 #include "gen_cpp/PlanNodes_types.h"
 #include "runtime/runtime_state.h"
+
 namespace starrocks {
 
+class ExprContext;
 namespace vectorized {
 class ConnectorScanNode;
-// class ConnectorSinkNode;
+class RuntimeFilterProbeCollector;
 } // namespace vectorized
 
 namespace connector {
 
 class DataSource {
 public:
-    virtual Status open(RuntimeState* state);
-    virtual void close(RuntimeState* state);
-    virtual Status get_next(RuntimeState* state, vectorized::ChunkPtr* chunk);
-
-    virtual Status do_open(RuntimeState* state) = 0;
-    virtual void do_close(RuntimeState* state) = 0;
-    virtual Status do_get_next(RuntimeState* state, vectorized::ChunkPtr* chunk) = 0;
+    virtual ~DataSource() = default;
+    virtual Status open(RuntimeState* state) { return Status::OK(); }
+    virtual void close(RuntimeState* state) {}
+    virtual Status get_next(RuntimeState* state, vectorized::ChunkPtr* chunk) { return Status::OK(); }
+
+    virtual int64_t raw_rows_read() const { return 0; }
+    virtual int64_t num_rows_read() const { return 0; }
+
+    void set_runtime_profile(RuntimeProfile* runtime_profile) { _runtime_profile = runtime_profile; }
+    void set_predicates(const std::vector<ExprContext*>& predicates) { _conjunct_ctxs = predicates; }
+    void set_runtime_filters(const vectorized::RuntimeFilterProbeCollector* runtime_filters) {
+        _runtime_filters = runtime_filters;
+    }
+    void set_read_limit(const uint64_t limit) { _read_limit = limit; }
+
+protected:
+    int64_t _read_limit = -1; // no limit
+    std::vector<ExprContext*> _conjunct_ctxs;
+    const vectorized::RuntimeFilterProbeCollector* _runtime_filters;
+    RuntimeProfile* _runtime_profile;
 };
 
 using DataSourcePtr = std::unique_ptr<DataSource>;
 
 class DataSourceProvider {
 public:
-    virtual Status init(RuntimeState* state) = 0;
+    virtual ~DataSourceProvider() = default;
+
     // First version we use TScanRange to define scan range
     // Later version we could use user-defined data.
     virtual DataSourcePtr create_data_source(const TScanRange& scan_range) = 0;
```

Comment:
why creating date source only need TScanRange?

Reply:
You could think `DataSource` is very similar to `ChunkSource`. It  reads a chunk of data in that scan range.

Find granularity will be supported in future because olap scan node will suport split chunk sources in a single scan range.